src/EventSubscriber/ElasticApmRequestSubscriber.php

```diff
@@ -0,0 +1,238 @@
+   *   The request event object.
+   */
+  public function onRequest(GetResponseEvent $event) {
+    // Don't process if Elastic APM is not configured.
+    if (!$this->elasticApm->isConfigured()) {
+      return;
+    }
+
+    // If this is a sub request, only process it if there was no master
+    // request yet. In that case, it is probably a page not found or access
+    // denied page.
+    if ($event->getRequestType() !== HttpKernelInterface::MASTER_REQUEST && $this->processedMasterRequest) {
+      return;
+    }
+
+    // Start a new transaction.
+    try {
+      $transaction = $this->phpAgent->startTransaction($this->routeMatch->getRouteName());
+
+      // Capture database time by wrapping spans around the db queries run.
+      $transaction->setSpans($this->constructDatabaseSpans());
+    }
+    catch (Exception $e) {
+      // Notify the user of the error.
+      $this->messenger->addError(t('An error occurred while trying to send the transaction to the Elastic APM server.'));
+
+      // Log the error to watchdog.
+      $error = Error::decodeException($e);
+      $this->logger->log($error['severity_level'], '%type: @message in %function (line %line of %file).', $error);
+    }
+
+    // Mark the request as processed.
+    $this->processedMasterRequest = TRUE;
+  }
+
+  /**
+   * End the transaction and send to PHP Agent whenever this event is triggered.
+   *
+   * @param \Symfony\Component\HttpKernel\Event\FinishRequestEvent $event
+   *   The event to process.
+   */
+  public function onFinishRequest(FinishRequestEvent $event) {
+    // Don't process if Elastic APM is not configured.
+    if (!$this->elasticApm->isConfigured()) {
+      return;
+    }
+
+    // Don't process if we don't have a PHP Agent already initialized, meaning,
+    // no transaction is in process.
+    if (!$this->phpAgent) {
+      return;
+    }
+
+    // End the transaction.
+    try {
+      $this->phpAgent->stopTransaction($this->routeMatch->getRouteName());
+
+      // Send our transaction to Elastic.
+      $this->phpAgent->send();
+    }
+    catch (Exception $e) {
+      // Notify the user of the error.
+      $this->messenger->addError(t('An error occurred while trying to send the transaction to the Elastic APM server.'));
+
+      // Log the error to watchdog.
+      $error = Error::decodeException($e);
+      $this->logger->log($error['severity_level'], '%type: @message in %function (line %line of %file).', $error);
+    }
+  }
+
+  /**
+   * Create spans around the db queries that were run in this request.
+   *
+   * @return array
+   *   An array of spans that will be added to the transaction.
+   */
+  protected function constructDatabaseSpans() {
+    $spans = [];
+
+    // First, go through all queries that have been run for this request.
+    $connections = [];
+    foreach (Database::getAllConnectionInfo() as $key => $info) {
+      $database = Database::getConnection('default', $key);
```

Comment:
**CHANGE REQUEST**

*Testability*

Database service should be injected via constructor.

Reply:
@krystalcode Drupal\Core\Database\Database is not a service. See class description:

>  * This class is uninstantiatable and un-extendable. It acts to encapsulate
>  * all control and shepherding of database connections into a single location
>  * without the use of globals.